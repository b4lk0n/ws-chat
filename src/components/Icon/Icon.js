import React from 'react';
import { node, oneOf, oneOfType, string } from 'prop-types';
import { isString } from 'lodash-es';
import cx from 'classnames';

import styles from './Icon.module.css';

export const Icon = ({ value, size, className }) => {
  return (
    <span
      className={cx(
        styles.root,
        styles[size],
        {
          'material-icons': isString(value),
        },
        className
      )}
    >
      {value}
    </span>
  );
};
Icon.propTypes = {
  value: oneOfType([string, node]).isRequired,
  size: oneOf(['small', 'large']),
  className: string,
};
