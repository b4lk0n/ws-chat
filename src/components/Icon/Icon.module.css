.root {
  --size: 24px;

  display: inline-flex;
  justify-content: center;
  align-items: center;
  font-size: var(--size);
  line-height: 1;
  width: 1em;
  height: 1em;
}

.small {
  --size: 18px;
}

.large {
  --size: 36px;
}
