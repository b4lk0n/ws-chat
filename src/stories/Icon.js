import React from 'react';
import { withKnobs, text, select } from '@storybook/addon-knobs';
import { Icon } from '../components';

export default {
  title: 'Icon',
  decorators: [withKnobs],
};

export const Default = () => (
  <Icon
    value={text('value', 'search')}
    size={select('size', {
      'â€”': undefined,
      small: 'small',
      large: 'large',
    })}
  />
);
