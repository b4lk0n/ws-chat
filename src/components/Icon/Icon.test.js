import React from 'react';
import { render, getNodeText } from '@testing-library/react';
import { Icon } from './Icon';

const CustomIcon = () => <span />;

describe('<Icon />', () => {
  it('should render children', () => {
    const { container } = render(<Icon value={'account'} />);
    expect(getNodeText(container.firstChild)).toBe('account');
  });

  it('should not add additional classes by default', () => {
    const { container } = render(<Icon value={'account'} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should add additional classes if provided', () => {
    const { container } = render(<Icon value={'account'} className={'colored'} />);
    expect(container.firstChild).toHaveClass('colored');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not add material-icons class for custom icon elements', () => {
    const { container } = render(<Icon value={<CustomIcon />} />);
    expect(container.firstChild).not.toHaveClass('material-icons');
    expect(container.firstChild).toMatchSnapshot();
  });

  ['small', 'large'].forEach(size => {
    it(`should render ${size} icon`, () => {
      const { container } = render(<Icon value={'account'} size={size} />);
      expect(container.firstChild).toHaveClass(size);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
