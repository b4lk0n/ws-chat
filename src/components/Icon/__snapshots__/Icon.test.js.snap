// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should add additional classes if provided 1`] = `
<span
  class="root undefined material-icons colored"
>
  account
</span>
`;

exports[`<Icon /> should not add additional classes by default 1`] = `
<span
  class="root undefined material-icons"
>
  account
</span>
`;

exports[`<Icon /> should not add material-icons class for custom icon elements 1`] = `
<span
  class="root undefined"
>
  <span />
</span>
`;

exports[`<Icon /> should render large icon 1`] = `
<span
  class="root large material-icons"
>
  account
</span>
`;

exports[`<Icon /> should render small icon 1`] = `
<span
  class="root small material-icons"
>
  account
</span>
`;
